theano/sandbox/cuda/dnn.py

```diff
@@ -2184,8 +2184,406 @@ def method(self):
+
+    Parameters
+    ----------
+    mode
+        Normalization is performed per-activation for 'per-activation' (in this mode bnBias and bnScale tensor
+        dimensions are 1xCxHxW for 4d tensor) or Normalization is performed over N+spatial dimensions for 'spatial'
+        (in this mode bnBias, bnScale tensor dimensions are 1xCx1x1).
+    epsilon
+        Epsilon value used in the batch normalization formula. It is 1e-4 by default. Minimum allowed value
+        for cuDNN v.4 is 1e-5.
+
+    """
+
+    direction = 'inference'
+    input_descs = [('bn_input', 0), ('scale_bias_mean', 1)]
+
+    def make_node(self, x, bn_scale, bn_bias, estimated_mean, estimated_variance):
+        x = as_cuda_ndarray_variable(x)
+        bn_scale = as_cuda_ndarray_variable(bn_scale)
+        bn_bias = as_cuda_ndarray_variable(bn_bias)
+        estimated_mean = as_cuda_ndarray_variable(estimated_mean)
+        estimated_variance = as_cuda_ndarray_variable(estimated_variance)
+        assert x.ndim == 4
+        assert bn_scale.ndim == 4
+        assert bn_bias.ndim == 4
+        assert estimated_mean.ndim == 4
+        assert estimated_variance.ndim == 4
+        return Apply(self, [x, bn_scale, bn_bias, estimated_mean, estimated_variance],
+                     [x.type()])
+
+    def method(self):
+        return """
+#ifndef CUDNN_VERSION
+err%(name)s = cudnnBatchNormalizationForwardInference(
+  _handle,
+  mode%(name)s,
+  bn_input_%(name)s,
+  CudaNdarray_DEV_DATA(%(ins0)s),
+  bn_output_%(name)s,
+  CudaNdarray_DEV_DATA(%(outs0)s),
+  scale_bias_mean_%(name)s,
+  CudaNdarray_DEV_DATA(%(ins1)s),
+  CudaNdarray_DEV_DATA(%(ins2)s),
+  CudaNdarray_DEV_DATA(%(ins3)s),
+  CudaNdarray_DEV_DATA(%(ins4)s),
+  epsilon%(name)s
+);
+#else
+{
+const float alpha = 1.;
+const float beta = 0.;
+err%(name)s = cudnnBatchNormalizationForwardInference(
+  _handle,
+  mode%(name)s,
+  (void*) &alpha,
+  (void*) &beta,
+  bn_input_%(name)s,
+  CudaNdarray_DEV_DATA(%(ins0)s),
+  bn_output_%(name)s,
+  CudaNdarray_DEV_DATA(%(outs0)s),
+  scale_bias_mean_%(name)s,
+  CudaNdarray_DEV_DATA(%(ins1)s),
+  CudaNdarray_DEV_DATA(%(ins2)s),
+  CudaNdarray_DEV_DATA(%(ins3)s),
+  CudaNdarray_DEV_DATA(%(ins4)s),
+  epsilon%(name)s
+);
+}
+#endif
+"""
+
+
+class GpuDnnBatchNorm(GpuDnnBatchNormBase):
+    """
+    Op for the cuDNN BatchNormalizationForwardTraining
+
+    Parameters
+    ----------
+    mode
+        Normalization is performed per-activation for 'per-activation' (in this mode bnBias and bnScale tensor
+        dimensions are 1xCxHxW for 4d tensor) or Normalization is performed over N+spatial dimensions for 'spatial'
+        (in this mode bnBias, bnScale tensor dimensions are 1xCx1x1).
+    epsilon
+        Epsilon value used in the batch normalization formula. It is 1e-4 by default. Minimum allowed value
+        for cuDNN v.4 is 1e-5.
+
+    """
+
+    direction = 'forward'
+    input_descs = [('bn_input', 0), ('scale_bias_mean', 1)]
+    destroy_map = {1: [3], 2: [4]}
```

Comment:
Also, the inplace parameter influence the c code. For the not inplace case, we should copy the input that are being worked inplace by cudnn instead of calling *prep_output()